### PR TITLE
PLATY-337 Support for checking file size/type constraints. Storing callback urk as a metadata.

### PR DESCRIPTION
**Pull request by tomasz-rosiek**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,15 +14,27 @@ should contain details about expected upload, which include additional metadata,
 about content type and size, and callback URL which will be used to notify user. Here is an example of the request:
 ```
 {
-	"callbackUrl": "http://myservice.com/callback?fileId=123"
+	"callbackUrl": "http://myservice.com/callback?fileId=123",
+	"minimumFileSize" : 0,
+	"maximumFileSize" : 1024,
+	"expectedMimeType": "application/xml"
```
Comment (konradwudkowski) [inline README.md]: is there only one possible mime type per upload?

Comment (tomasz-rosiek) [reply README.md]: Well - :) That's the limitation of AWS :)

File: README.md [hunk 2]
```diff
@@ -14,15 +14,27 @@ should contain details about expected upload, which include additional metadata,
...
+|maximumFileSize|Maximum file size, if not specified, global maximum file size will be applied (by default 100MB)|no|
+|expectedMimeType|Expected MIME type of uploaded file|no|
+
 The service replies with JSON containg reference of the upload and information about the POST form that has to be sent in order to upload the file:
 ```
 {
     "reference": "11370e18-6e24-453e-b45a-76d3e32ea33d",
     "uploadRequest": {
-        "href": "https://bucketName.s3.eu-west-2.amazonaws.com",
+        "href": "https://bucketName.s3.eu-west-2.amazsonaws.com",
```
Comment (konradwudkowski) [inline README.md]: fat fingers?

File: README.md (hunk 1 above)
Comment (konradwudkowski) [reply README.md]: what is the size unit? Bytes? Can you specify with prefix e.g. 100M, or 10KB?

File: app/controllers/PrepareUploadController.scala [hunk 3]
```diff
@@ -3,35 +3,42 @@ package controllers
...
-  implicit val uploadedFileSettingsFormat = Json.format[UploadSettings]
-
-  implicit val linkWrites: Writes[Link] = (
-    (JsPath \ "href").write[String] and
-      (JsPath \ "method").write[String]
-  )(unlift(Link.unapply))
+  implicit val uploadSettingsReads: Reads[UploadSettings] = (
+    (JsPath \ "callbackUrl").read[String] and
+      (JsPath \ "minimumFileSize").readNullable[Int](min(0)) and
+      (JsPath \ "maximumFileSize").readNullable[Int](min(0) keepAnd max(prepareUploadService.globalFileSizeLimit + 1)) and
```
Comment (konradwudkowski) [inline app/controllers/PrepareUploadController.scala]: why `+1` in the max size?

Comment (tomasz-rosiek) [reply app/controllers/PrepareUploadController.scala]: Reads.max is not inclusive.

File: README.md (hunk 1 above)
Comment (tomasz-rosiek) [reply README.md]: Integer, bytes

File: test/controllers/PrepareUploadControllerSpec.scala [hunk 4]
```diff
@@ -22,33 +22,61 @@ class PrepareUploadControllerSpec extends UnitSpec with Matchers with GivenWhenT
...
+            "maximumFileSize" -> 1024))
 
       When("upload initiation has been requested")
 
       val result = controller.prepareUpload()(request)
 
       Then("service returns valid response with reference and template of upload form")
 
-      status(result) shouldBe 200
+      withClue(result) { status(result) shouldBe 200 }
```
Comment (konradwudkowski) [inline test/controllers/PrepareUploadControllerSpec.scala]: if result is a Future wouldn't the message in `withClue` look like `Promise@13aaa14a`

File: test/infrastructure/s3/S3PrepareUploadServiceSpec.scala [hunk 5]
```diff
@@ -40,21 +51,115 @@ class S3PrepareUploadServiceSpec extends UnitSpec with Matchers with GivenWhenTh
...
+          expectedContentType = None)
+
+      When("we setup the upload")
+      Then("an exception should be thrown")
+
+      an[IllegalArgumentException] shouldBe thrownBy(service.setupUpload(uploadSettings))
+
+    }
+
+    "fail when minimum file size is greater than minimum file size" in {
```
Comment (konradwudkowski) [inline test/infrastructure/s3/S3PrepareUploadServiceSpec.scala]: greater than **maximum** file size

File: test/infrastructure/s3/S3PrepareUploadServiceSpec.scala [hunk 6]
```diff
@@ -40,21 +51,115 @@ class S3PrepareUploadServiceSpec extends UnitSpec with Matchers with GivenWhenTh
...
+        UploadSettings(
+          callbackUrl         = callbackUrl,
+          minimumFileSize     = Some(0),
+          maximumFileSize     = Some(1025),
+          expectedContentType = None)
+
+      When("we setup the upload")
+      Then("an exception should be thrown")
+
+      an[IllegalArgumentException] shouldBe thrownBy(service.setupUpload(uploadSettings))
```
Comment (konradwudkowski) [inline test/infrastructure/s3/S3PrepareUploadServiceSpec.scala]: would it make sense to assert on the message or throw strongly type exceptions?

File: README.md (hunk 1 above)
Comment (konradwudkowski) [reply README.md]: would it make sense to document it in the readme?

File: app/controllers/PrepareUploadController.scala (hunk 3 above)
Comment (konradwudkowski) [reply app/controllers/PrepareUploadController.scala]: :+1:

**Pull request merged**